examples/ecdh.c

```diff
@@ -0,0 +1,77 @@
+#include <stdio.h>
+#include <assert.h>
+#include <string.h>
+
+#include "random.h"
+#include "secp256k1.h"
+#include "secp256k1_ecdh.h"
+
+
+void print_hex(unsigned char* data, size_t size) {
+    size_t i;
+    printf("0x");
+    for (i = 0; i < size; i++) {
+        printf("%02x", data[i]);
+    }
+    printf("\n");
+}
+
+int main(void) {
+    unsigned char seckey1[32];
+    unsigned char seckey2[32];
+    unsigned char compressed_pubkey1[33];
+    unsigned char compressed_pubkey2[33];
+    unsigned char shared_secret1[32];
+    unsigned char shared_secret2[32];
+    size_t len;
+    secp256k1_pubkey pubkey1;
+    secp256k1_pubkey pubkey2;
+    secp256k1_context* ctx = secp256k1_context_create(SECP256K1_CONTEXT_SIGN);
+    while (1) {
+        if (!fill_random(seckey1, sizeof(seckey1)) || !fill_random(seckey2, sizeof(seckey2))) {
+            printf("Failed to generate randomness\n");
+                return 1;
+        }
+        if (secp256k1_ec_seckey_verify(ctx, seckey1) && secp256k1_ec_seckey_verify(ctx, seckey2)) {
+            break;
+        }
+    }
+
+    /* Pubkey creation on a valid Context with a verified seckey should never fail */
+    assert(secp256k1_ec_pubkey_create(ctx, &pubkey1, seckey1));
+    assert(secp256k1_ec_pubkey_create(ctx, &pubkey2, seckey2));
+
+    /* Serialize the pubkey in a compressed form */
+    len = sizeof(compressed_pubkey1);
+    secp256k1_ec_pubkey_serialize(ctx, compressed_pubkey1, &len, &pubkey1, SECP256K1_EC_COMPRESSED);
+    /* Should be the same size as the size of the output */
+    assert(len == sizeof(compressed_pubkey1));
+
+    /* Serialize the pubkey in a compressed form */
+    len = sizeof(compressed_pubkey2);
+    secp256k1_ec_pubkey_serialize(ctx, compressed_pubkey2, &len, &pubkey2, SECP256K1_EC_COMPRESSED);
+    assert(len == sizeof(compressed_pubkey2));
+
+    /* Perform ECDH with seckey1 and pubkey2, should never fail with a verified seckey and valid pubkey */
+    assert(secp256k1_ecdh(ctx, shared_secret1, &pubkey2, seckey1, NULL, NULL));
+
+    /* Perform ECDH with seckey2 and pubkey1, should never fail with a verified seckey and valid pubkey */
+    assert(secp256k1_ecdh(ctx, shared_secret2, &pubkey1, seckey2, NULL, NULL));
+
+    /* Both parties should end up with the same shared secret */
+    assert(memcmp(shared_secret1, shared_secret2, sizeof(shared_secret1)) == 0);
+
+    printf("Secret Key1: ");
+    print_hex(seckey1, sizeof(seckey1));
+    printf("Compressed Pubkey1: ");
+    print_hex(compressed_pubkey1, sizeof(compressed_pubkey1));
+    printf("\nSecret Key2: ");
+    print_hex(seckey2, sizeof(seckey2));
+    printf("Compressed Pubkey2: ");
+    print_hex(compressed_pubkey2, sizeof(compressed_pubkey2));
+    printf("\nShared Secret: ");
+    print_hex(shared_secret1, sizeof(shared_secret1));
+
+
```

Comment:
We need to destroy the context here. Also can try to clear the secret keys (and add comment why we do that).

Reply:
Weird that valgrind didn't complain about it.

Reply:
Should I do try to do some best effort in that regard? (ie #636)

Reply:
Shall I export a function to clear memory in #636?

Reply:
I think it's an improvement, although probably most serious users will have their own mechanism.

Reply:
Done, I added memset with a TODO, so we could either copy your function to the example, or if we actually export it use that.